### PR TITLE
[Tensorflow] [Inference]TF 2.9.3 release

### PR DESCRIPTION
**Pull request by hballuru**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/sanity/test_pre_release.py [hunk 1]
```diff
@@ -184,6 +184,53 @@ def test_tf_serving_version_cpu(tensorflow_inference):
 
     stop_and_remove_container(container_name, ctx)
 
+# Added by hballuru to confirm tf-serving-api version is matching tf serving version
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_pre_release.py:187]: Please remove the contributor name from the comment. Git blame can help one do that

File: test/dlc_tests/sanity/test_pre_release.py [hunk 2]
```diff
@@ -184,6 +184,53 @@ def test_tf_serving_version_cpu(tensorflow_inference):
...
+
+    if "gpu" in image:
+        pytest.skip(
+            "GPU images will have their framework version tested in test_framework_and_cuda_version_gpu")
+    if "neuron" in image:
+        pytest.skip(
+            "Neuron images will have their framework version tested in test_framework_and_neuron_sdk_version")
+
+    _, tag_framework_version = get_framework_and_version_from_tag(
+        image)
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_pre_release.py:216]: nit: Is there any reason got having it in next line?

File: test/dlc_tests/sanity/test_pre_release.py [hunk 3]
```diff
@@ -184,6 +184,53 @@ def test_tf_serving_version_cpu(tensorflow_inference):
 
     stop_and_remove_container(container_name, ctx)
 
+# Added by hballuru to confirm tf-serving-api version is matching tf serving version
+@pytest.mark.usefixtures("sagemaker")
+@pytest.mark.model("N/A")
+@pytest.mark.canary("Run non-gpu tf serving version test regularly on production images")
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_pre_release.py:190]: Do you see any specific reason for running the test on canary?

File: test/dlc_tests/sanity/test_pre_release.py [hunk 4]
```diff
@@ -184,6 +184,53 @@ def test_tf_serving_version_cpu(tensorflow_inference):
...
+
+    @param tensorflow_inference: ECR image URI
+    """
+    # Set local variable to clarify contents of fixture
+    image = tensorflow_inference
+
+    if "gpu" in image:
+        pytest.skip(
+            "GPU images will have their framework version tested in test_framework_and_cuda_version_gpu")
+    if "neuron" in image:
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_pre_release.py:211]: We haven't added anything to the neuron image function. Do we need to add that? Or, we can run this test for `Neuron` as well?

File: test/dlc_tests/sanity/test_pre_release.py [hunk 5]
```diff
@@ -184,6 +184,53 @@ def test_tf_serving_version_cpu(tensorflow_inference):
...
+    container_name = get_container_name("tf-serving-api-version", image)
+    start_container(container_name, image, ctx)
+    output = run_cmd_on_container(
+        container_name, ctx, "pip show tensorflow-serving-api | grep Version", executable="bash"
+    )
+    str_version_from_output = ((str(output.stdout).split(' '))[1]).strip()
+    assert (tag_framework_version == str_version_from_output), \
+        f"Tensorflow serving API version is {str_version_from_output} while the Tensorflow version is {tag_framework_version}. Both don't match!"
+
+    stop_and_remove_container(container_name, ctx)
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_pre_release.py:233]: Can we use a `try` and `final` strategy to ensure that `stop_and_remove_container` is executed even if the assert statement fails?

File: test/dlc_tests/sanity/test_pre_release.py [hunk 6]
```diff
@@ -184,6 +184,53 @@ def test_tf_serving_version_cpu(tensorflow_inference):
...
+    to be spun up.
+
+    @param tensorflow_inference: ECR image URI
+    """
+    # Set local variable to clarify contents of fixture
+    image = tensorflow_inference
+
+    if "gpu" in image:
+        pytest.skip(
+            "GPU images will have their framework version tested in test_framework_and_cuda_version_gpu")
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_pre_release.py:210]: I read the comment above saying `This test can run directly on the host, and thus does not require additional resources to be spun up.` . If that is the case, why do we want to tackle GPU test separately. Why can't we allow this test to run for GPU images as well. I think that this test does not spawn any ec2 instance and it would be judicious if we use the same test for GPU images.

File: test/dlc_tests/sanity/test_pre_release.py (hunk 4 above)
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py]: The test case has been redesigned to handle neuron

File: test/dlc_tests/sanity/test_pre_release.py (hunk 2 above)
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py]: fixed

File: test/dlc_tests/sanity/test_pre_release.py (hunk 1 above)
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py]: Done

File: test/dlc_tests/sanity/test_pre_release.py (hunk 3 above)
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py]: removed canary

File: test/dlc_tests/sanity/test_pre_release.py (hunk 6 above)
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py]: the code has been moved away from the method

File: test/dlc_tests/sanity/test_pre_release.py (hunk 5 above)
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py]: try catch finally has been implemented to ensure container removal. However bear in mind there are other parts of the same test where the container removal is not enclosed within try block.

File: test/dlc_tests/sanity/test_pre_release.py [hunk 7]
```diff
@@ -184,6 +184,43 @@ def test_tf_serving_version_cpu(tensorflow_inference):
...
+        output = run_cmd_on_container(
+            container_name, ctx, cmd, executable="bash"
+        )
+        str_version_from_output = ((str(output.stdout).split(' '))[1]).strip()
+        assert (tag_framework_version == str_version_from_output), \
+            f"Tensorflow serving API version is {str_version_from_output} while the Tensorflow version is {tag_framework_version}. Both don't match!"
+    except Exception as e:
+        LOGGER.error(f"Unable to execute command on container. Error: {e}")
+        raise        
+    finally:
```
Comment (kenny-ezirim) [inline test/dlc_tests/sanity/test_pre_release.py:222]: Let's remove finally here except if start_container is going to be moved into the try-except block. Also, verify what happens if we cannot start a container with the name. Does the command "start_container" return a value? We can use that value to decide if we need to raise an exception or stop the container after running the test command. 

The assumptions in this piece of code are 1. container name is always given and the containers are always instantiated successfully.

File: test/dlc_tests/sanity/test_pre_release.py [hunk 8]
```diff
@@ -184,6 +184,43 @@ def test_tf_serving_version_cpu(tensorflow_inference):
...
+    @param tensorflow_inference: ECR image URI
+    """
+    # Set local variable to clarify contents of fixture
+    image = tensorflow_inference
+
+    if "gpu" in image:
+        cmd="pip show tensorflow-serving-api-gpu | grep Version"
+    if "cpu" in image:
+        cmd="pip show tensorflow-serving-api | grep Version"
+
```
[the remark this answers is not in the document]
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py:206]: We really don't anticipate a new image that will miss the check above. Only thing we can think is "hpu" but that doesn't come under inference so we are good. For now I did go ahead and added an else block to handle the edge case if it may arise. Thanks for the suggestion

File: test/dlc_tests/sanity/test_pre_release.py (hunk 7 above)
Comment (hballuru) [reply test/dlc_tests/sanity/test_pre_release.py:222]: The container doesnt need to be inside try block as per the review. The start container can fail by itself and that has nothing to do with finally. The stop and remove container logic only executes when something in the try block fails because we know that the code within try gets executed only upon successful start of container and we want a cleanup/removal of such.

**Review by shantanutrip — DISMISSED** [review]

**Comment by shantanutrip** [comment]
Please revert the temp changes

**Review by shantanutrip — APPROVED** [review]
